main/src/library/Map.flix

```diff
@@ -1,7 +1,266 @@
 namespace Map {
 
-    def empty[k, v]: Map[k, v] = ???
+    //
+    // ## Map Construction
+    //
 
-    def insert[k, v](k: k, v: v, m: Map[k, v]): Map[k, v] = ???
+    ///
+    /// Returns the empty map.
+    ///
+    def empty[k, v]: Map[k, v] = Map(Nil)
 
+    ///
+    /// Returns the singleton map where key `k` is mapped to value `v`.
+    ///
+    def singleton[k, v](k: k, v: v): Map[k, v] = Map((k, v) :: Nil)
+
+    //
+    // ## Basic Operations
+    //
+
+    ///
+    /// Returns `true` if and only if `m` is the empty map, i.e. `Map(Nil)`.
+    ///
+    def null[k, v](m: Map[k, v]): Bool =
+        let Map(xs) = m;
+            List.isEmpty(xs)
+
+    ///
+    /// Returns `Some(v)` if `k -> v` is in `m`.
+    /// Otherwise returns `None`.
+    ///
+    def get[k, v](k: k, m: Map[k, v]): Option[v] = match m with {
+        case Map(Nil) => None
+        case Map((a, b) :: rs) => if (a == k) Some(b) else get(k, Map(rs))
+    }
+
+    ///
+    /// Returns `v` if `k -> v` is in `m`.
+    /// Otherwise, returns `d`.
+    ///
+    def getWithDefault[k, v](k: k, d: v, m: Map[k, v]): v = match m with {
+        case Map(Nil) => d
+        case Map((a, b) :: rs) => if (a == k) b else getWithDefault(k, d, Map(rs))
+    }
+
+    ///
+    /// Returns `true` if and only if `m` contains the key `k`.
+    ///
+    def memberOf[k, v](k: k, m: Map[k, v]): Bool = match m with {
+        case Map(Nil) => false
+        case Map((a, b) :: rs) => if (a == k) true else memberOf(k, Map(rs))
+    }
+
+    ///
+    /// Returns the keys of `m`.
+    ///
+    def keysOf[k, v](m: Map[k, v]): Set[k] =
+        let Map(xs) = m;
+            Set(fst(List.unzip(xs)))
```

Comment:
Beautiful!

A minor thing, you should not use the `Set` constructor here (inside the implementation of Map). The reason is, we cannot make any assumptions about Set here. Instead, you just need to call `List.toSet(...)`. This way, if we ever change the implementation of set it won't affect map.